architecture/ambient/lifecycle.md

```diff
@@ -0,0 +1,136 @@
+# Ztunnel Lifecyle
+
+This document describes the lifecycle of various aspects of Ztunnel, and its relationship to workloads and the CNI.
+
+At a high level, our goal is to provide complete connectivity to a workload throughout its entire lifetime.
+Failing to do so can be an availability risk (if we deny traffic that should succeed) or a security risk (if we allow traffic that should be denied).
+
+## High level overview
+
+At a high level, the relevant components look as such:
+
+```mermaid
+flowchart TD
+    CNIP["CNI Plugin"]
+    CNI["CNI Agent"]
+    Z["Ztunnel"]
+    Pods["Pods"]
+    CNI --Programs--> Z
+    CNI --Watches--> Pods
+    Z --Runs Within--> Pods
+    CNIP --Calls--> CNI
+```
+
+The CNI Plugin is a binary installed as a [CNI plugin](https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/network-plugins/) on the node.
+Kubernetes is responsible for invoking this when a Pod is being started (before the containers run).
+When this occurs, the plugin will call out to the CNI Agent to program the network.
+This includes setting up networking rules both within the pod network namespace, and also on the host network.
+
+An alternative flow is when a pod is enabled after startup.
+In this case, the CNI Agent is watching for Pod events from the API server directly and performing the same setup.
+
+Once the network is configured, the CNI Agent will signal to Ztunnel to start running within the Pod.
+This is done by the [ZDS](../../pkg/zdsapi/zds.proto) API.
+This will send some identifying information about the Pod to Ztunnel, and, importantly, the Pod's network namespace file descriptor.
+
+Ztunnel will use this to enter the Pod network namespace and start various listeners (inbound, outbound, etc).
+
+> [!NOTE]
+> While Ztunnel runs as a single shared binary on the node, each individual pod gets its own unique set of listeners within its own network namespace.
+
+## Pod Startup
+
+### Pod Startup Requirements
+
+During Pod startup, we MUST meet the constraint that the network is ready the instant the application starts (including `initContainers`!).
+This is critical -- issues around this were the [top issue with sidecars](https://github.com/istio/istio/issues/11130).
+
+### Pod Startup Implementation
+
+The key mechanism to maintain this property is the CNI plugin.
+The CNI plugin synchronously runs before application containers start, ensure we can setup the network configuration and start Ztunnel.
+Only once those two conditions are met do we mark the CNI plugin as successfully complete.
```

Comment:
What are the error cases ? What happens if ztunnel pod is not deployed ? 

My understanding is that CNI pod will keep in memory a list with all the network namespaces, and send them to ztunnel if it shows up ( and if CNI daemon is restarted - will do the /proc scanning to guess )

Reply:
It blocks the pod startup. I updated the doc